### DISCUSSION TITLE
Auto populate/display/update field value based on another field using table data

### DISCUSSION THREAD

**Discussion by nunick** [Budibase/budibase#4145] [Help] [answered]
Is it possible to auto populate/display/update a value of a field based on the contents of another when they share the same row in a table? For example, say I enter in my name into a name field and then the email field below automatically populates/displays my email from a table where both values are stores.

EDIT: I can get a Headline field to auto populate to the same value as selected in an Option Picker on the screen using something like: {{ Form.Fields.Option }} or similar. However, I'd like it to be the email associated with the user, not the user. And instead of an Option Picker and Header, it'd be useful to be able to type any string into one field and then have it populate another.

![image](https://user-images.githubusercontent.com/94874984/150615081-bb65e65a-1ff0-4e56-84ea-2b86d7dedfd4.png)

**Comment by mjashanks** [comment]
Hi @nunick you can use the `Default Value` field to do this...

![image](https://user-images.githubusercontent.com/3524181/150766786-e4bfddfd-01a0-42c4-9167-a1c612e44649.png)

My binding in this image is `{{ Form.Fields.Name }} - {{ Form.Fields.Company }}`.

So, when I complete the `Name` and `Company` fields in the form below, the `Description` is auto-populated. You can of course override this.

![image](https://user-images.githubusercontent.com/3524181/150767021-45c93943-d27f-4214-8d93-7154127c886e.png)

**Reply by nunick** [reply]
Hi @mjashanks, thank you for the example! I am able to replicate that behaviour on my end. Unfortunately, I'm not able create the situation where (using your example) a user types (or scans) the name "Mike" and then in the "Description" field "Budibase" pops up automatically. The field "Company" wouldn't exist on the page.

Essentially the field "Description" would get the value "Company" from the table row where "Name" "Mike" is stored. Not sure if the syntax exists, but I'm imagining something like {{ Form.Fields.Name.Company }} or {{ Form.Fields.Name(Company) }}.

A behaviour like this would be useful if say I have a profile for a product (serial number, part number, description, cost, etc) and I need to create a maintenance report for a product a customer returns. I would enter/scan the serial number for the return product and the screen would pull a few fields from the product profile and enter it into the maintenance report automatically.

Another example would be similar to autocomplete when filling out Shipping Details when ordering something. Usually you type your name in the field and the browser asks if you want to fill in the rest of the details associated with that profile.

**Reply by mjashanks** [reply]
This is also possible - there are a few ways to do it, depending on what you need to do. In my video below, I have a setup where
1. The "Org" text box is used to filter a dataprovider
2. The "Repo" options picker uses the dataprovider as a source (so it's items are filtered based on the "Org" text
3. The Default value for selected repo is set based on the selection in "Repo"

You could also 
1. type to list entries in a table.
2. add a "Select" button in the table to select a row. It would use a "Set State" action to store the selected row data
3. You could bind the rest of the form's default value to the data you have just stored in state

https://user-images.githubusercontent.com/3524181/151075296-3dfe57d0-dc3f-423a-b63f-ab6769346030.mp4

**Reply by nunick** [reply]
Thank you for the follow-up @mjashanks! I'm trying option 1 as it looks like what I want to do. However, I don't see the option for "Get Repo" or the settings gear for my Data Provider to add a Query Binding. I'm currently running 1.0.46 self hosted. Am I using the wrong component?

![image](https://user-images.githubusercontent.com/94874984/151223765-0400e092-53f2-4d1b-a808-fbc1b89d2ccd.png)

![image](https://user-images.githubusercontent.com/94874984/151223792-a19cb271-ca2d-47e2-802f-6560761fdad7.png)

**Reply by mjashanks** [reply]
Ah yea, sorry - I was using a REST API data provider, and my query was named "Get Repo". The Settings cog comes with the REST API.

You can do the same with internal tables, or SQL by using "Define Filters" instead

**Reply by nunick** [reply]
@mjashanks update: I was able to figure it out! Not sure if I implemented it totally correctly, but it's behaving how I want.

I was able to set up a Postgres db with two tables: `maintenance_report` and `serial_number_master` with a query to the serial table. The query grabs the serial number, part number, and product description.

![image](https://user-images.githubusercontent.com/94874984/151436784-62baebca-5bc5-4746-93fe-3e68d3bf1e63.png)
![image](https://user-images.githubusercontent.com/94874984/151596900-f04550de-366e-4d82-8099-d5e5205853c6.png)

On my screen I linked the Data Provider with the query and bound results based on the serial number entered. Entering a serial number will filter the query and the Product Number and Product Description fields populate with the data called from the row "0" of the query: `{{ New Data Provider.Rows.0.product_description }}`

![image](https://user-images.githubusercontent.com/94874984/151597680-5ebe0a43-f9b6-40e9-acc2-27e3ae1f12fc.png)
![image](https://user-images.githubusercontent.com/94874984/151597965-5af88e44-66bc-413f-b6be-0f0f7b9447fd.png)

With the forms auto populated, I add the additional information in the summary and then the Save button adds the information to the "maintenance_report" table.

![image](https://user-images.githubusercontent.com/94874984/151598045-76474f22-f73d-4d86-98bf-d2956d45867d.png)

One thing I'm not too clear on is I had to make the form Schema the Query as well, otherwise a bunch of forms were reading "This Field setting is the wrong data type for this component." <- Still figuring out the component structure on the platform.

**Reply by mjashanks** [reply]
Ah - awesome - thanks for reporting your experience back! This will be super helpful to others :)

**Reply by Deluzionist** [reply]
I have a filter binding based on a form field on a data provider that links to a MsSQL query. Entering a known value from the table brings back no results. 
![filter example](https://user-images.githubusercontent.com/46769667/152251139-d8406458-9735-48c8-9f0f-c542eabec2fc.PNG)

**Reply by nunick** [reply]
@Deluzionist in my example that I got working I didn't bind anything under `Filtering` -> `Define Filters`. Instead my bind was set when I created the query and adding bindings under then query and then on my screen in the `Data Provider` I clicked the gear beside the `Data` source and then added the field to the set binding under the query. See my previous message from Jan 28.

**Reply by mjashanks** [reply]
Thanks @nunick - that's correct.

When using a custom query....
1. The gear will pass parameters into your query
2. The "Define Filters" will apply a filter to the results **Client Side / In the Browser**

When using tables, "Define Filters" will apply the filters on the backend - i.e. directly to SQL.

**Reply by Deluzionist** [reply]
Thanks @nunick.

@mjashanks Is there any way to use a form field to pull values from a different data provider from the initial form schema?
Right now I have two forms with two different data providers. One pulls query information to populate the field default values on the other form used to populate another existing table. Either I can't pull data onto the second form with the handlebar expressions or the form isn't updating the selected table.

**Reply by mjashanks** [reply]
@Deluzionist - not sure what you mean - could you give me some screenshot to show what you are trying to do?

**Reply by Deluzionist** [reply]
@mjashanks It's hard to show the connections in a screenshot but let's say I have table A and table B that shares a few column values. I would like to use a form which will update table B using values filtered from table A.
This requires two forms, one to filter table A and the second to update table B. The universal handlebar pathway is not binding form fields B to values from form fields A.

**Reply by nunick** [reply]
@Deluzionist in my example I'm taking things from Table A (serial number table) that the user filters with a field, add a few new details in a summary field, and then have a save button to save the info from all fields into the form to Table B (maintenance report). This is how I've set up my button:

![image](https://user-images.githubusercontent.com/94874984/152612799-7e11486f-fbfc-4883-a8b6-6135522efadf.png)

That being said, I haven't set it up to "update" the cells in Table B if needs, but that's probably best for a different screen to handle.

**Reply by Deluzionist** [reply]
@nunick Thanks for elucidating. I have tried using just one form as you have done but when I use the handlebar expression {{ dataprovider.Rows.0.columnname }} it will only pull up values if the dataprovider matches the schema chosen for the form.

**Reply by mjashanks** [reply]
Does this help?

https://user-images.githubusercontent.com/3524181/152631592-864c52fa-6e43-40a2-869a-e9f5c82e3e4e.mp4

**Reply by Deluzionist** [reply]
@mjashanks Thank you! The JS works great however I still can't get the Data Provider filtering to work the way you had implemented the binding  {{ Fields.Code }} via Define Filters. The only way I was able to get the filtering to work was by binding the data source query itself the way nunick had implemented via the where column = handlebar expression.
I had even replicated the filtering from https://docs.budibase.com/guides/add-a-dropdown-filter-to-a-table exactly. Any idea what I could be doing wrong?

**Reply by mjashanks** [reply]
Happy to take a look at your app, if you want to send to support@budibase.com

**Comment by melohagan** [comment, marked as answer]
Using the default value is the old approach. Default values only apply on load of the screen.

If you want to update one field with the value of another, look at the [On Change event](https://docs.budibase.com/docs/text-inputs#on-change) in combination with the [Update Field Value action](https://docs.budibase.com/docs/form-actions#update-field-value)
